docarray/array/mixins/attribute.py

```diff
@@ -21,6 +23,11 @@ def _get_array_attribute(
         in the array like container
         """
         field_type = self.__class__.document_type._get_nested_document_class(field)
+        if is_optional_type(field_type):
+            field_type = get_args(field_type)[0]  # get type inside Optional
+        elif is_union_type(field_type):
+            # determine type based on the first element
+            field_type = type(getattr(self[0], field))
```

Comment:
so this looks at the data to determine the type right ? Smart idea

Reply:
exactly, I don't like it so much actually, but I think it's the only way